ui/app/pages/confirm-approve/confirm-approve.util.js

```diff
@@ -1,28 +1,27 @@
 import { decimalToHex } from '../../helpers/utils/conversions.util'
 import { calcTokenValue } from '../../helpers/utils/token-util.js'
+import { getTokenData } from '../../helpers/utils/transactions.util'
 
-export function getCustomTxParamsData (data, { customPermissionAmount, tokenAmount, decimals }) {
-  if (customPermissionAmount) {
-    const tokenValue = decimalToHex(calcTokenValue(tokenAmount, decimals))
+export function getCustomTxParamsData (data, { customPermissionAmount, decimals }) {
+  const tokenData = getTokenData(data)
 
-    const re = new RegExp('(^.+)' + tokenValue + '$')
-    const matches = re.exec(data)
-
-    if (!matches || !matches[1]) {
-      return data
-    }
-    let dataWithoutCurrentAmount = matches[1]
-    const customPermissionValue = decimalToHex(calcTokenValue(Number(customPermissionAmount), decimals))
-
-    const differenceInLengths = customPermissionValue.length - tokenValue.length
-    const zeroModifier = dataWithoutCurrentAmount.length - differenceInLengths
-    if (differenceInLengths > 0) {
-      dataWithoutCurrentAmount = dataWithoutCurrentAmount.slice(0, zeroModifier)
-    } else if (differenceInLengths < 0) {
-      dataWithoutCurrentAmount = dataWithoutCurrentAmount.padEnd(zeroModifier, 0)
-    }
+  if (!tokenData) {
+    throw new Error(`Invalid data`)
+  } else if (tokenData.name !== 'approve') {
+    throw new Error(`Invalid data; should have be 'approve' method, but instead is '${tokenData.name}'`)
+  }
+  let spender = tokenData.params[0].value
+  if (spender.startsWith('0x')) {
+    spender = spender.substring(2)
+  }
+  const [signature, tokenValue] = data.split(spender)
 
-    const customTxParamsData = dataWithoutCurrentAmount + customPermissionValue
-    return customTxParamsData
+  let customPermissionValue = decimalToHex(calcTokenValue(customPermissionAmount, decimals))
+  if (customPermissionValue.length > tokenValue.length) {
```

Comment:
Why `customPermissionValue.length > tokenValue.length`?

`tokenValue` is the approve amount set by the dapp correct? If `tokenValue` < `customPermissionValue` < [largest possible token value], wouldn't this exclude some valid `customPermissionValue`s?

Reply:
Oh, wait, I get it

Reply:
Actually, I don't. Why can't we have a `customPermissionValue` of length greater than `tokenValue` length?

Reply:
`tokenValue` here is the zero-padded token value, which is always of length 64 (which is how much space a `u256` takes up in hex). I should probably explain this in the code - and maybe assert my expectations then check for greater-than-64 explicitly 🤔

Reply:
> and maybe assert my expectations then check for greater-than-64 explicitly
yes, and if so, we can do that validation earlier in the code

Reply:
I'm not entirely sure what you mean; where do you imagine this validation should take place? I've updated this - let me know what you think